execute/tokendata/observer.go

```diff
@@ -2,42 +2,146 @@ package tokendata
 // Goal of that is to support multiple token observers supporting different tokens (e.g. CCTP, MyFancyToken etc)
-type CompositeTokenDataObserver struct {
-	Observers []TokenDataObserver
+type compositeTokenDataObserver struct {
+	observers []TokenDataObserver
+}
+
+// nolint early-return unexported-return
+func NewConfigBasedCompositeObservers(config []pluginconfig.TokenDataObserverConfig) (*compositeTokenDataObserver, error) {
+	observers := make([]TokenDataObserver, len(config))
+	for i, c := range config {
+		if c.USDCCCTPObserverConfig != nil {
+			observers[i] = usdc.NewTokenDataObserver(*c.USDCCCTPObserverConfig, nil, nil)
+		} else {
+			return nil, errors.New("unsupported token data observer")
+		}
+	}
+	return NewCompositeObservers(observers...), nil
+}
+
+// nolint unexported-return
+func NewCompositeObservers(observers ...TokenDataObserver) *compositeTokenDataObserver {
+	return &compositeTokenDataObserver{observers: observers}
 }
 
-func (t *CompositeTokenDataObserver) Observe(
+// Observe start with stubbing exectypes.TokenDataObservations with empty data based on the supported tokens.
+// Then it iterates over all observers and merges token data returned from them into the final result.
+func (t *compositeTokenDataObserver) Observe(
 	ctx context.Context,
-	messages exectypes.MessageObservations,
+	msgObservations exectypes.MessageObservations,
 ) (exectypes.TokenDataObservations, error) {
-	var tokenObservation exectypes.TokenDataObservations
-	for _, ob := range t.Observers {
-		tokenData, err := ob.Observe(ctx, messages)
+	tokenDataObservations := t.initTokenDataObservations(msgObservations)
+
+	for _, ob := range t.observers {
+		tokenData, err := ob.Observe(ctx, msgObservations)
+		if err != nil {
+			return nil, err
+		}
+		tokenDataObservations, err = merge(tokenDataObservations, tokenData)
 		if err != nil {
 			return nil, err
 		}
-		tokenObservation = merge(tokenObservation, tokenData)
 	}
-	return tokenObservation, nil
+	return tokenDataObservations, nil
 }
 
-func merge(_ exectypes.TokenDataObservations, _ exectypes.TokenDataObservations) exectypes.TokenDataObservations {
-	return exectypes.TokenDataObservations{}
+func (t *compositeTokenDataObserver) IsTokenSupported(
+	chainSelector cciptypes.ChainSelector,
+	token cciptypes.RampTokenAmount,
+) bool {
+	for _, ob := range t.observers {
+		if ob.IsTokenSupported(chainSelector, token) {
+			return true
+		}
+	}
+	return false
+}
+
+// initTokenDataObservations initializes the token data observations with empty token data, it asks the child observers
+// whether the token is supported or not. If token is supported and requires additional processing we set its state to
+// isReady=false. If token is noop (doesn't require offchain processing) we initialize it with empty
+// data and set IsReady=true.
+func (t *compositeTokenDataObserver) initTokenDataObservations(
```

Comment:
`getSupportedTokenDataObs` better?

Reply:
`getSupportedTokenDataObs` that's not true, we initialize everything, but set proper values based on whether the token is supported or not

Reply:
ahaa... one generic comment on the approach is why use slices and check for lengths and have noop/unsupported? Isn't it better to use a map of messageID to tokenData when only it has data and return the tokenData mapped by messageID?